### PR TITLE
Migrate rustc_resolve to use SessionDiagnostic, part # 1

### PR DESCRIPTION
**Pull request by rajputrajat**
crate a somewhat on larger size, so plz allow some time to get it finished.

### CODE REVIEW

**Comment by compiler-errors** [comment]
@rajputrajat, can you please remove all of these merge commits? We have a no merge commit policy.

**Comment by compiler-errors** [comment]
@rajputrajat  -- I think you need to do something like `git pull --rebase $REMOTE master` where `$REMOTE` is the git remote corresponding to upstream, i.e. the rust-lang/rust repo (not your fork).

**Comment by rajputrajat** [comment]
Yes. I will update this as you suggested.
Being a first timer, I did read this comment somewhere on some other PR but didn't understand it then.

Thanks!

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 1]
```diff
@@ -798,6 +738,9 @@ impl<'a> Resolver<'a> {
                         (span_with_rename.shrink_to_lo(), "{".to_string()),
                         (span_with_rename.shrink_to_hi(), "}".to_string()),
                     ];
+                    // rajputrajat: multipart_suggestion is not yet supported.
+                    // will implement when its available.
+                    // leave it for now
```
Comment (davidtwco) [inline compiler/rustc_resolve/src/diagnostics.rs:743]: cc #100970

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 2]
```diff
@@ -843,27 +771,16 @@ impl<'a> Resolver<'a> {
...
-                    self.session,
+                self.session.create_err(errs::CannotCaptureDynamicEnvironmentInFnItem {
                     span,
-                    E0434,
-                    "{}",
-                    "can't capture dynamic environment in a fn item"
-                );
-                err.help("use the `|| { ... }` closure form instead");
-                err
+                    closure_hint: "|| { ... }",
```
Comment (davidtwco) [inline compiler/rustc_resolve/src/diagnostics.rs:776]: Why can't this be in the Fluent message?

File: compiler/rustc_resolve/src/errors.rs [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+#[diag(resolve::crate_root_imports_must_be_named_explicitly)]
+pub(crate) struct ResolutionError(#[primary_span] pub(crate) Span);
+
+#[derive(SessionDiagnostic)]
+#[diag(resolve::name_is_already_used_as_generic_parameter, code = "E0403")]
+pub(crate) struct NameAlreadyUsedInParameterList {
+    #[primary_span]
+    #[label]
+    pub(crate) span: Span,
+
```
Comment (davidtwco) [inline compiler/rustc_resolve/src/errors.rs:39]: We can remove the newlines between fields :)

File: compiler/rustc_resolve/src/errors.rs [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    pub(crate) ident: Ident,
+    pub(crate) suggestion: &'a str,
+    pub(crate) current: &'a str,
+}
+
+#[derive(SessionSubdiagnostic)]
+#[label(resolve::attempt_to_use_non_constant_value_in_constant_with_suggestion_label)]
+pub(crate) struct AttemptToUseNonConstantValueInConstantWithSuggestionLabel {
```
Comment (davidtwco) [inline compiler/rustc_resolve/src/errors.rs:267]: I think you've got some of these types mixed up with the Fluent messages, it looks like this message expects a `suggestion` field that it doesn't have but that `AttemptToUseNonConstantValueInConstantWithoutSuggestion` does (despite its name).

**Review by davidtwco — COMMENTED** [review]
This is a great start, I've left some comments, you'll need to remove the merge commits as @compiler-errors has noted. :)

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 2 above)
Comment (rajputrajat) [reply compiler/rustc_resolve/src/diagnostics.rs:775]: fluent parser was failing on this closure string for some reason. I think it was due to '{'. I would need to look into in again.

File: compiler/rustc_resolve/src/errors.rs (hunk 3 above)
Comment (rajputrajat) [reply compiler/rustc_resolve/src/errors.rs:39]: sure, thing.

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 2 above)
Comment (rajputrajat) [reply compiler/rustc_resolve/src/diagnostics.rs:775]: Similar thing is failing to parse at another place.
image
For now I am passing just code related info from struct field.
Please let me know if there is something which could be done about it. Maybe I am missing something.

File: compiler/rustc_resolve/src/errors.rs (hunk 4 above)
Comment (rajputrajat) [reply compiler/rustc_resolve/src/errors.rs]: tried to rename this type to more meaningful and understandable as per context.

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_resolve/src/diagnostics.rs:762]: Use `{"{"}` to escape `{`, and similarly for `}`. https://projectfluent.org/fluent/guide/special.html

**Comment by RalfJung** [comment]
Looks like something went wrong in the rebase, this contains a bunch of changes it shouldn't (submodules, lockfile).

File: compiler/rustc_error_messages/locales/en-US/resolve.ftl [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+resolve_binding_shadows_something_unacceptable_suggestion =
+    try specify the pattern arguments
+
+resolve_forward_declared_generic_param =
+    generic parameters with a default cannot use forward declared identifiers
+    .label = defaulted generic parameters cannot be forward declared
+
+resolve_param_in_ty_of_const_param =
+    the type of const parameters must not depend on other generic parameters
+    .label = the type must not depend on the parameter `{name}`
```
Comment (davidtwco) [inline compiler/rustc_error_messages/locales/en-US/resolve.ftl:146]:
```suggestion
    .label = the type must not depend on the parameter `{$name}`
```
Shouldn't it be this? Similarly with all other uses of variables in the Fluent resource.

Comment (rajputrajat) [reply compiler/rustc_error_messages/locales/en-US/resolve.ftl:146]: silly me. will fix it all as soon as possible.

**Comment by davidtwco** [comment]
You should run `./x.py test src/test/ui` locally and try to resolve the test failures as you go - so that the diagnostics you are migrating continue to work :)

**Comment by rajputrajat** [comment]
> You should run `./x.py test src/test/ui` locally and try to resolve the test failures as you go - so that the diagnostics you are migrating continue to work :)

yep. working on the failed tests.

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 6]
```diff
@@ -703,128 +667,78 @@ impl<'a> Resolver<'a> {
...
-
-                err
+                    None => ((None, None), None),
+                };
+                self.session.create_err(errs::UndeclaredLabel {
+                    span,
+                    name,
+                    sub_reachable,
+                    sub_reachable_suggestion,
+                    sub_unreachable,
```
Comment (davidtwco) [inline compiler/rustc_resolve/src/diagnostics.rs:711]: Would using an enum for these subdiagnostics be better? They seem mutually exclusive.

File: compiler/rustc_resolve/src/errors.rs [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+    symbol::{Ident, Symbol},
+    Span,
+};
+
+#[derive(Diagnostic)]
+#[diag(resolve::parent_module_reset_for_binding, code = "E0637")]
+pub(crate) struct ParentModuleResetForBinding;
+
+#[derive(Diagnostic)]
+#[note]
```
Comment (davidtwco) [inline compiler/rustc_resolve/src/errors.rs:12]: nit: could you put the `#[diag]` attribute first?

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 8]
```diff
@@ -897,135 +812,78 @@ impl<'a> Resolver<'a> {
...
-                err.span_label(
-                    span,
-                    format!("cannot be named the same as {} {}", article, shadowed_binding_descr),
-                );
-                match (shadowing_binding, shadowed_binding) {
+            } => self.session.create_err(errs::BindingShadowsSomethingUnacceptable {
+                span,
+                shadowing_binding_descr: shadowing_binding.descr(),
+                shadowed_binding_descr: shadowed_binding.descr(),
+                article,
```
Comment (davidtwco) [inline compiler/rustc_resolve/src/diagnostics.rs:819]: Is this going to be translatable? Things like `article` are probably quite English-specific?

File: compiler/rustc_error_messages/locales/en-US/resolve.ftl [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    try using a local generic parameter instead
+
+resolve_try_adding_local_generic_param_on_method =
+    try adding a local generic parameter in this method instead
+
+resolve_help_try_using_local_generic_param =
+    try using a local generic paramter instead
+
+resolve_name_is_already_used_as_generic_parameter =
+    the name `{ $name }` is already used for a generic parameter in this item's generic parameters
```
[the remark this answers is not in the document]
Comment (rajputrajat) [reply compiler/rustc_error_messages/locales/en-US/resolve.ftl:48]: I was trying to follow the convention as suggested here:
https://projectfluent.org/fluent/guide/variables.html

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 6 above)
Comment (rajputrajat) [reply compiler/rustc_resolve/src/diagnostics.rs:714]: They definitely seems so. However, on closer inspection you may notice that, in the case when `suggestion` == `Some(_, True)`, then `label` and `suggestion` both needs to be displayed, and an enum variant can only let us handle one diagnostic at a time.
Please suggest if there is some way to handle such scenario.

**Comment by davidtwco** [comment]
Are you still migrating this or is this ready to be landed?

**Comment by rajputrajat** [comment]
> Are you still migrating things or is this ready to be landed?

I am still in the process. There is still a lot of stuff to migrate. Is it possible to merge this migration in parts to minimise the review scope?

**Comment by davidtwco** [comment]
> > Are you still migrating things or is this ready to be landed?
> 
> I am still in the process. There is still a lot of stuff to migrate. Is it possible to merge this migration in parts to minimise the review scope?

We can merge this in parts if you want.

**Comment by Nilstrieb** [comment]
Hi, I've seen you changed some diagnostic structs in your PR. After #103345, the way we refer to fluent messages changed. They are now in a flat namespace with the same identifier as in the fluent file. For example,  `parser::cool_thing` is now `parser_cool_thing` and `parser::suggestion` just `suggestion`.
You should [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts) to the latest master and change your fluent message references as described above. Thanks!

**Comment by rajputrajat** [comment]
> Hi, I've seen you changed some diagnostic structs in your PR. After #103345, the way we refer to fluent messages changed. They are now in a flat namespace with the same identifier as in the fluent file. For example, `parser::cool_thing` is now `parser_cool_thing` and `parser::suggestion` just `suggestion`. You should [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts) to the latest master and change your fluent message references as described above. Thanks!

made changes as suggeste. Thanks!

**Comment by rajputrajat** [comment]
> > > Are you still migrating things or is this ready to be landed?
> > 
> > 
> > I am still in the process. There is still a lot of stuff to migrate. Is it possible to merge this migration in parts to minimise the review scope?
> 
> We can merge this in parts if you want.

Could you initiate merging for this one, if possible, please?
There is still one comment opened with this one though, regarding "article" and "participle" being only English specific. But that can be opened separately.
I will open separate PR, named as part#2 for newer changes.

Thanks!

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 10]
```diff
@@ -1039,6 +899,17 @@ impl<'a> Resolver<'a> {
...
+                //     span,
+                //     name,
+                //     kind: kind.to_owned(),
+                //     trait_item_span,
+                //     trait_path,
+                //     code: match code {
+                //         DiagnosticId::Error(e) => e,
+                //         DiagnosticId::Lint { name, .. } => name,
+                //     },
+                // })
```
Comment (davidtwco) [inline compiler/rustc_resolve/src/diagnostics.rs:912]: Could you remove this commented code?

**Review by davidtwco — COMMENTED** [review]
This looks good to me, apologies for the delay in reviewing, just one minor comment - would appreciate if you could squash some of the commits too.

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 10 above)
Comment (rajputrajat) [reply compiler/rustc_resolve/src/diagnostics.rs]: Done.

**Comment by rajputrajat** [comment]
> This looks good to me, apologies for the delay in reviewing, just one minor comment - would appreciate if you could squash some of the commits too.

Done. Please review.

**Comment by rajputrajat** [comment]
> Some changes occurred in src/librustdoc/clean/types.rs
> 
> cc @camelid
> 
> Hey! It looks like you've submitted a new PR for the library teams!
> 
> If this PR contains changes to any `rust-lang/rust` public library APIs then please comment with `@rustbot label +T-libs-api -T-libs` to tag it appropriately. If this PR contains changes to any unstable APIs please edit the PR description to add a link to the relevant [API Change Proposal](https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html) or [create one](https://github.com/rust-lang/libs-team/issues/new?assignees=&labels=api-change-proposal%2C+T-libs-api&template=api-change-proposal.md&title=%28My+API+Change+Proposal%29) if you haven't already. If you're unsure where your change falls no worries, just leave it as is and the reviewer will take a look and make a decision to forward on if necessary.
> 
> Examples of `T-libs-api` changes:
> 
> * Stabilizing library features
> * Introducing insta-stable changes such as new implementations of existing stable traits on existing stable types
> * Introducing new or changing existing unstable library APIs (excluding permanently unstable features / features without a tracking issue)
> * Changing public documentation in ways that create new stability guarantees
> * Changing observable runtime behavior of library APIs

Wrong commits were picked while squashing. corrected!

**Review by davidtwco — CHANGES_REQUESTED** [review]

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 11]
```diff
@@ -922,135 +835,78 @@ impl<'a> Resolver<'a> {
...
-                    self.session,
+                let ((sub_suggestion_label, sub_suggestion), sub_unreachable_label) =
+                    match suggestion {
+                        // A reachable label with a similar name exists.
+                        Some((ident, true)) => (
+                            (
+                                Some(errs::UnreachableLabelSubLabel { ident_span: ident.span }),
+                                Some(errs::UnreachableLabelSubSuggestion {
+                                    span,
+                                    ident_name: ident.name,
```
[the remark this answers is not in the document]
Comment (rajputrajat) [reply compiler/rustc_resolve/src/diagnostics.rs:887]: this one being used in 'suggestion', so seems like need to leave as is.

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 12]
```diff
@@ -1099,48 +940,29 @@ impl<'a> Resolver<'a> {
...
-                );
-                err.span_suggestion(
-                    path.span,
-                    "try",
-                    format!("crate::{}", pprust::path_to_string(&path)),
-                    Applicability::MaybeIncorrect,
-                );
-                err
+                    path_span: path.span,
+                    path_str: pprust::path_to_string(&path),
```
[the remark this answers is not in the document]
Comment (rajputrajat) [reply compiler/rustc_resolve/src/diagnostics.rs:946]: that string is part of a suggestion, so need to leave as is.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**